handler/core/src/main/java/io/knotx/fragments/handler/action/cache/CacheAction.java

```diff
@@ -0,0 +1,120 @@
+import io.knotx.fragments.handler.api.domain.FragmentContext;
+import io.knotx.fragments.handler.api.domain.FragmentResult;
+import io.knotx.server.api.context.ClientRequest;
+import io.knotx.server.common.placeholders.PlaceholdersResolver;
+import io.knotx.server.common.placeholders.SourceDefinitions;
+import io.vertx.core.AsyncResult;
+import io.vertx.core.Future;
+import io.vertx.core.Handler;
+import io.vertx.core.json.JsonObject;
+import io.vertx.core.logging.Logger;
+import io.vertx.core.logging.LoggerFactory;
+import org.apache.commons.lang3.StringUtils;
+
+public class CacheAction implements Action {
+
+  private static final Logger LOGGER = LoggerFactory.getLogger(CacheAction.class);
+
+  private final Action doAction;
+
+  private final String payloadKey;
+
+  private final JsonObject config;
+
+  private final Cache cache;
+
+  public CacheAction(Action doAction, JsonObject config,
+      Cache cache) {
+    this.doAction = doAction;
+    this.config = config;
+    this.payloadKey = getPayloadKey(config);
+    this.cache = cache;
+  }
+
+  @Override
+  public void apply(FragmentContext fragmentContext, Handler<AsyncResult<FragmentResult>> resultHandler) {
+    String cacheKey = getCacheKey(config, fragmentContext.getClientRequest());
+    cache.get(cacheKey)
+        .subscribe(cachedValue -> {
+          LOGGER.trace("Cache responsed with value: {}", cachedValue);
+          Fragment fragment = fragmentContext.getFragment();
+          fragment.appendPayload(payloadKey, cachedValue);
+          FragmentResult result = new FragmentResult(fragment, FragmentResult.SUCCESS_TRANSITION);
+          Future.succeededFuture(result)
+              .setHandler(resultHandler);
+        }, error -> {
+          LOGGER.error("Could not fetch data from cache. {}", error);
+          Fragment fragment = fragmentContext.getFragment();
+          FragmentResult result = new FragmentResult(fragment, FragmentResult.ERROR_TRANSITION);
+          Future.succeededFuture(result)
+              .setHandler(resultHandler);
+        }, () -> {
+          LOGGER.trace("Cache didn't return value.");
+          callDoActionAndCache(fragmentContext, resultHandler, cacheKey);
+        });
+  }
+
+  private void callDoActionAndCache(FragmentContext fragmentContext,
+      Handler<AsyncResult<FragmentResult>> resultHandler, String cacheKey) {
+    doAction.apply(fragmentContext, asyncResult -> {
+      if (asyncResult.succeeded()) {
+        FragmentResult fragmentResult = asyncResult.result();
+        if (FragmentResult.SUCCESS_TRANSITION.equals(fragmentResult.getTransition())
+            && fragmentResult.getFragment()
+            .getPayload()
+            .containsKey(payloadKey)) {
+          JsonObject resultPayload = fragmentResult.getFragment()
+              .getPayload();
+          cache.put(cacheKey, resultPayload.getMap()
+              .get(payloadKey));
+        }
+        Future.succeededFuture(fragmentResult)
```

Comment:
consider whether this should be configurable

Reply:
what if `doAction` specifies a `_fallback` transition, but there is something that could be cached?
In general: what is the approach for handling decorated actions' results/transitions